crates/sui-core/src/authority_active/checkpoint_executor/mod.rs

```diff
@@ -0,0 +1,236 @@
+                    // will attempt to execute all checkpoints from (highest_executed_checkpoint + 1)
+                    // up to the result of the next recv() call.
+                    Err(RecvError::Lagged(num_skipped)) => {
+                        warn!(
+                            "Checkpoint Execution Recv channel overflowed {:?} messages",
+                            num_skipped,
+                        );
+                        self.metrics
+                            .checkpoint_exec_recv_channel_overflow
+                            .inc_by(num_skipped);
+                    }
+                    Err(RecvError::Closed) => {
+                        info!("Checkpoint Execution Sender (StateSync) closed channel");
+                        break;
+                    }
+                },
+            }
+        }
+    }
+
+    async fn reconfig(&self, _next_epoch_committee: &[(AuthorityName, StakeUnit)]) -> SuiResult {
+        // TODO(william) call into Reconfig
+        Ok(())
+    }
+
+    fn update_lag_metric(&self, highest_executed: CheckpointSequenceNumber) -> SuiResult {
+        let highest_synced = self
+            .checkpoint_store
+            .get_highest_synced_checkpoint_seq_number()?
+            .unwrap_or(0);
+        // TODO may want to consider asserting >= 0
+        let diff = highest_synced - highest_executed;
+        self.metrics.checkpoint_exec_lag.set(diff as i64);
+        Ok(())
+    }
+}
+
+pub async fn execute_checkpoint(
+    checkpoint: VerifiedCheckpoint,
+    authority_state: Arc<AuthorityState>,
+    checkpoint_store: Arc<CheckpointStore>,
+) -> SuiResult<Vec<TransactionEffects>> {
+    let effects = if let Some(checkpoint_contents) = checkpoint_store
+        .get_checkpoint_contents(&checkpoint.content_digest())
+        .map_err(SuiError::from)?
+    {
+        let txes = checkpoint_contents.into_inner();
+        execute_and_verify_transactions(txes, authority_state).await?
+    } else {
+        warn!(
+            "Checkpoint contents empty: {:?}",
+            checkpoint.content_digest()
+        );
+        Vec::new()
+    };
+    Ok(effects)
+}
+
+async fn execute_and_verify_transactions(
+    transactions: Vec<ExecutionDigests>,
+    authority_state: Arc<AuthorityState>,
+) -> SuiResult<Vec<TransactionEffects>> {
+    let tx_digests: Vec<TransactionDigest> = transactions.iter().map(|tx| tx.transaction).collect();
+
+    // TODO Need to acquire shared object locks currently. Instead, replcae with call to transaction manager
+    // using transaction effects (which has object version numbers and therefore is already sequenced)
+    let mut verified_certs = Vec::<VerifiedCertificate>::new();
+    for digest in tx_digests.iter() {
+        match authority_state.database.read_certificate(digest)? {
+            Some(cert) => verified_certs.push(cert),
+            None => return Err(SuiError::TransactionNotFound { digest: *digest }),
+        }
+    }
```

Comment:
I believe you can call `multi_get_certified_transaction`

Reply:
This should be happening either in consensus or `StateSync` depending on the source. cc @bmwill to confirm

Reply:
`StateSync` when downloading transactions, would only insert them into pending_certificates, not certificates

Reply:
What is the difference between a `TrustedCertificate` (the form persisted in the `pending_certificates` table) and `VerifiedCertificate`? Do both these not represent a txn cert signed by a quorum that has not yet been executed locally?

Reply:
Because `TransactionManager` accepts the latter. But it seems `StateSync` persists the former.

Reply:
You can look at the comments in the definition of `TrustedEnvelope`. It's conceptually the same as Verified except it's for db store.

Reply:
Ah I see the fix now. Thanks for pointing this out